File: src/findCommonKmers.py
'''

	Author - Mayank Pahadia
	Program - Find the common kmers between two kmercount files. 

	Run the program - python findCommonKmers.py input1.txt input2.txt


'''

import sys
from collections import defaultdict

if(len(sys.argv)<5):
	print "Please pass the input files"
	print "Example - python findCommonKmers.py tempKmerOut.txt input1.txt input2.txt commonKmersFrom1in2.txt commonKmersFrom2in1.txt"
	exit()

tempKmerOut=sys.argv[1]
input1 = sys.argv[2]
inputs = sys.argv[3:]

kmers = defaultdict(lambda: 0)

print input1, inputs


countInput1=0

delim="\t"


out = open(tempKmerOut,'w')
c=""
c="Percentage of Kmers from Reference File in Input File" + delim + "Percentage of Kmers from Input File in Reference File" + "\n"
out.write(c)

with open(input1) as f:
	for line in f:
		line=line.replace('\n','')
		line=line.split()
		kmers[line[0]]=1
		countInput1+=1

c=''
ijk=0
print countInput1
while(ijk<len(inputs)):
	commonFrom1 = defaultdict(lambda: 0)
	count=0
	countInput2=0
	input2 = inputs[ijk]
	ijk+=1
	file12=open(inputs[ijk],'w')
	ijk+=1
	file21=open(inputs[ijk],'w')
	ijk+=1
	with open(input2) as f:
		for line in f:
			line=line.replace('\n','')
			countInput2+=1
			line=line.split()
			c=''
			c+=line[0] + delim + str(kmers[line[0]])
			c+='\n'
			file12.write(c)
			if(kmers[line[0]]==1):
				commonFrom1[line[0]]=1
				count+=1

	for key in kmers:
		c=''
		c+=str(key) + delim
		if commonFrom1[key]==1:
			c+=str(commonFrom1[key])
		else:
			c+=str(0)
		c+='\n'
		file21.write(c)

	# print "total kmers in", input1, " - ", countInput1
	# print "total kmers in", input2, " - ", countInput2
	# print "common kmers - ", count
	if(countInput1 == 0):
		percent1=0
	else:
		percent1 = count*1.0 / countInput1
		percent1*=100.0

	# print "common kmers from the total kmers in", input1, " - ", percent, "%"
	
	if(countInput2 == 0):
		percent2=0
	else:
		percent2 = count*1.0 / countInput2
		percent2*=100.0
	

	# print "common kmers from the total kmers in", input2, " - ", percent, "%"

	c=""
	c+=str(percent1) + "%" + delim + str(percent2) + "%" + "\n"
	out.write(c)

	file12.close()
	file21.close()

out.close()

File: src/findCommonKmers.cpp
#include <fstream>
#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;
int main(int argc, char* argv[])
{
	if(argc < 1){
		cout << "Provide inputfile as command line arguments\n";
		cout << "Example - ./a.out input1.txt input2.txt commonKmersFrom1in2.txt commonKmersFrom2in1.txt \n";
		exit(0);
	}
	
	string alignmentResults = argv[2];
	vector<string> inputFileNames,common12,common21;
    for(int i=3; i < argc; i++){
        inputFileNames.push_back(argv[i]);
        string temp = argv[i];
        string abc = ".commonKmers12";
        temp+=abc;
        common12.push_back(temp);
        temp = argv[i];
        abc = ".commonKmers21";
        temp+=abc;
        common21.push_back(temp);
    }
	string inputfile1 = argv[1], inputfile2;
	// cout << "here0\n";
	string outputfile1, outputfile2;
	string km;

	char delim[2]="\t";
	long long int value, count1=0, count2=0, commonKmers=0;

	unordered_map<string, int> kmers;
	unordered_map<string, int> kmersFrom1;
	ifstream input;
	ofstream output;
	// cout << "here1\n";

	input.open(inputfile1);
	if (input.is_open())
	{
		while(input >> km >> value){
			kmers[km]=1;
			count1++;
		}
	}
	input.close();


	ofstream alignRes(alignmentResults);
	string tempRes = "recall";
	tempRes+=delim;
	tempRes+= "precision";
	tempRes+= "\n";
	alignRes << tempRes;



	// cout << "here2\n";
	for(int ijk=0; ijk<inputFileNames.size();ijk++){

		inputfile2 = inputFileNames[ijk];
		outputfile1 = common12[ijk];
		outputfile2 = common21[ijk];
		commonKmers=0;
		count2=0;

		input.open(inputfile2);
		output.open(outputfile1);
		if (input.is_open())
		{
			while(input >> km >> value){
				kmersFrom1[km]=1;
				string c = km + delim;
				if(kmers[km]==1){
					commonKmers++;
					c+='1';
				}else{
					c+='0';
				}
				c+="\n";
				output << c;
				count2++;
			}
		}
		input.close();
		output.close();
		// cout << "here3\n";

		output.open(outputfile2);
		for(auto it=kmers.begin(); it!=kmers.end(); it++){
			string c=it->first + delim;
			if(kmersFrom1[it->first]==1){
				c+='1';
			}else{
				c+='0';
			}
			c+="\n";
			output << c;;
		}
		output.close();
		// cout << "here4\n";

		// cout << count1 << " " << count2 << " " << commonKmers << endl;
		cout << inputfile1 << " total kmers - " << count1 << endl;
		cout << inputfile2 << " total kmers - " << count2 << endl;
		cout <<"common kmers - " << commonKmers << endl;

		double percent = commonKmers*1.0 / count1;
		percent*=100.0;

		cout <<  "common kmers from the total kmers in " <<  inputfile1 << " - "<< percent << " % \n";

		alignRes << percent << "%" << delim;

		percent = commonKmers*1.0 / count2;
		percent*=100.0;

		cout <<  "common kmers from the total kmers in " <<  inputfile2 << " - "<< percent << " % \n";

		alignRes << percent << "%" << endl;

	}

	alignRes.close();





}